### PR TITLE
An initial Python attempt at splitting images

### PR DESCRIPTION
**Pull request by abkfenris**
In order to hopefully allow for faster image loading, 2i2c is letting us set up separate images for Python and R (and possible Python + ML libraries or other massive stuff).

I've started by creating a new directory for a Python image with a few updates to match the current 2i2c environment, and slimmed down further using micromamba. 

I've also created a [re-usable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) for building and pushing Docker images, that we can then reference from image/directory specific workflows.

I'm fully expecting needing to tinker on the workflows to get it building and pushing as expected.

### CODE REVIEW

**Comment by abkfenris** [comment]
I'm gonna merge without review so that I can debug and get 2i2c testing the images.

File: .github/workflows/built-image.yml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    name: Build and push image
+    timeout-minutes: 30
+
+    outputs:
+      image_name: ${{ steps.env_var.outputs.image_name }}
+      image_ref: ${{ steps.env_var.outputs.image_ref }}
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v2
```
Comment (ocefpaf) [inline .github/workflows/built-image.yml:63]:
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/built-image.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1.5.1
```
Comment (ocefpaf) [inline .github/workflows/built-image.yml:66]: The checkout one I do recommend the bump, this one I'm not sure.

```suggestion
        uses: docker/setup-buildx-action@v2
```

File: .github/workflows/built-image.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+      - name: Cache Docker layers
+        uses: actions/cache@v2.1.6
```
Comment (ocefpaf) [inline .github/workflows/built-image.yml:69]:
```suggestion
        uses: actions/cache@v3
```

File: .github/workflows/built-image.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          SHA7="${GITHUB_SHA::7}"
+          DOCKER_TAG=$SHA7
+          IMAGE_NAME="ghcr.io/${{ github.repository_owner }}/${{ inputs.image_name }}"
+          echo "DOCKER_TAG=${{ inputs.image_tag }}" >> $GITHUB_ENV
+          echo "IMAGE_NAME=${IMAGE_NAME}" >> $GITHUB_ENV
+          echo "::set-output name=image_name::${IMAGE_NAME})"
+          echo "::set-output name=image_ref::${DOCKER_TAG})"
+
+      - name: Build Docker Image
+        uses: docker/build-push-action@v2.6.1
```
Comment (ocefpaf) [inline .github/workflows/built-image.yml:88]:
```suggestion
        uses: docker/build-push-action@v3
```

File: .github/workflows/built-image.yml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+      - name: Export Full Conda Environment
+        run: |
+          docker run ${{ env.IMAGE_NAME }}:${{ DOCKER_TAG }} conda list --explicit > conda-packages.txt
+
+          echo "### Conda Environment" >> $GITHUB_STEP_SUMMARY
+          cat conda-packages.txt >> $GITHUB_STEP_SUMMARY
+
+      - name: Archive Conda Package List
+        uses: actions/upload-artifact@v1
```
Comment (ocefpaf) [inline .github/workflows/built-image.yml:113]:
```suggestion
        uses: actions/upload-artifact@v3
```

File: .github/workflows/built-image.yml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        with:
+          name: conda-packages
+          path: conda-packages.txt
+
+      - name: "Log into GitHub Container Registery"
+        uses: docker/login-action@v1.9.0
```
Comment (ocefpaf) [inline .github/workflows/built-image.yml:119]:
```suggestion
        uses: docker/login-action@v2
```

File: py-base/environment.yml [hunk 7]
```diff
@@ -0,0 +1,86 @@
+name: OHW-py
+channels:
+  - conda-forge
+dependencies:
+  - python=3.9
+  - pangeo-notebook=2022.07.12
```
Comment (ocefpaf) [inline py-base/environment.yml:6]: I'm pretty sure we can bump to 3.10 here. Fingers crosses!

```suggestion
  - python=3.10
  - pangeo-notebook=2022.07.12
```

**Review by ocefpaf — APPROVED** [review]
All minor comments, mostly GHA bumps. LGTM!

**Comment by abkfenris** [comment]
Thanks. I was going to keep dependency changes minimal at this stage, just so that we can get to the point of testing with 2i2c sooner. I've also had enough weirdness around Python 3.10 to leave that for a later PR.

Of course I also moved to Micromamba as a base at the same time, so my judgement may be lacking.

**Comment by ocefpaf** [comment]
Good strategy! For the GHA actions bumps we should probably adopt dependabot, not sure if we can enable it only for GHA. The Python 3.10 is not important. I'm mostly curious about what is broken.

**Comment by abkfenris** [comment]
I nearly wrote 'thanks, dependabot' You can enable it just for actions, but I would probably set it up for the Docker `FROM` images too.

Largely some weirdness about types.

**Pull request merged**